syntax = "proto3";

message DjiModuleInfo {
    string proto_name = 1;
    string unk1 = 2;
    string unk2 = 3;
    string serial_number = 5;
    string unk4 = 6;
    uint32 unk5 = 9;
    string camera_name = 10;
}

message DjiVideoGlobalInfo {
    repeated DjiModuleInfo module_info = 1;
}

message DjiRemoteGpsDeviceInfo {
    uint32 unk1 = 1;
    uint32 unk2 = 2;
    string device_name = 4;
    float unk3 = 5;
}

message DjiRemoteGpsCoordinatesInfo {
    uint32 coordinate_units = 1;
    double latitude = 2;
    double longitude = 3;
}

message CoordinatesDatetime {
    string datetime = 1;
}

message DjiRemoteGpsCoordinates {
    DjiRemoteGpsCoordinatesInfo info = 1;
    uint32 gps_altitude_mm = 2;
    uint32 unk2 = 4;
    uint32 unk3 = 5;
    CoordinatesDatetime datetime = 6;
}

message DjiRemoteDerivative {
    float x = 1;
    float y = 2;
    float z = 3;
}

message DjiRemoteGpsInfo {
    DjiRemoteGpsDeviceInfo device_info = 1;
    DjiRemoteGpsCoordinates coordinates = 2;
    DjiRemoteDerivative derivatives = 3;
}

message DjiUnkfloat_1 {
    float unknown = 1;
}

message DjiUnkInt32_1 {
    int32 unknown = 1;
}

message DjiUnkUInt32_1 {
    uint32 unknown = 1;
}

message DjiUnkRepBytes_1 {
    repeated bytes unknown_field = 1;
}

message DjiCameraAccelInfo {
    float unk = 1;
    float x = 2;
    float y = 3;
    float z = 4;
}

message DjiCameraInfo {
    DjiUnkfloat_1 unk1 = 2;
    DjiUnkfloat_1 unk2 = 3;
    DjiUnkRepBytes_1 unk3 = 4;
    DjiUnkfloat_1 unk4 = 5;
    DjiUnkUInt32_1 unk5 = 6;
    DjiUnkUInt32_1 unk6 = 7;
    uint32 unk7 = 8;
    DjiCameraAccelInfo accelerometer1 = 9;
    DjiCameraAccelInfo accelerometer2 = 10;
    uint32 unk8 = 11;
    uint32 unk9 = 12;
}

message GpsFrameInfo {
    uint32 frame_id = 1;
    uint32 unknown = 2;
}

message DjiGpsInfo {
    GpsFrameInfo frame_info = 1;
    DjiCameraInfo camera_info = 2;
    DjiRemoteGpsInfo remote_gps_info = 4;
}

message VideoStreamInfoDetails {
    uint32 width = 1;
    uint32 height = 2;
    float frame_rate = 3;
    uint32 unk1 = 4;
    uint32 unk2 = 5;
    uint32 unk3 = 6;
    uint32 unk5 = 8;
}

message VideoStreamInfo {
    string type = 1;
    VideoStreamInfoDetails details = 3;
}

message GenericMessage {
    DjiVideoGlobalInfo          video_global_info = 1;
    VideoStreamInfo             video_stream_info = 2;
    repeated DjiGpsInfo   gps_info = 3;
}
